File: Google/522.longest-uncommon-subsequence-ii.java
/*
 * @lc app=leetcode id=522 lang=java
 *
 * [522] Longest Uncommon Subsequence II
 *
 * Test candidate strings from longest to shortest. A candidate is uncommon if
 * it is not a subsequence of any other string with length at least as large.
 * The first such candidate has maximum length.
 *
 * Time: O(n^2 * L). Space: O(n).
 */

import java.util.Arrays;

// @lc code=start
class Solution {
    public int findLUSlength(String[] strs) {
        Integer[] order = new Integer[strs.length];
        for (int i = 0; i < strs.length; i++) {
            order[i] = i;
        }
        Arrays.sort(order, (a, b) -> Integer.compare(strs[b].length(), strs[a].length()));

        for (int index : order) {
            String candidate = strs[index];
            boolean uncommon = true;
            for (int other = 0; other < strs.length; other++) {
                if (other == index) {
                    continue;
                }
                if (strs[other].length() >= candidate.length() && isSubsequence(candidate, strs[other])) {
                    uncommon = false;
                    break;
                }
            }
            if (uncommon) {
                return candidate.length();
            }
        }
        return -1;
    }

    private boolean isSubsequence(String small, String large) {
        int pointer = 0;
        for (int i = 0; i < large.length() && pointer < small.length(); i++) {
            if (small.charAt(pointer) == large.charAt(i)) {
                pointer++;
            }
        }
        return pointer == small.length();
    }
}
// @lc code=end

File: companies/Bloomberg/522.longest-uncommon-subsequence-ii.go
package leetcode

//
// @lc app=leetcode id=522 lang=golang
//
// [522] Longest Uncommon Subsequence II
//
// Notes
// Test candidate strings from longest to shortest. A string is uncommon if it
// is not a subsequence of any other string with length at least as large. The
// first uncommon candidate in descending length order is optimal. Time:
// O(n^2 * L). Space: O(n) for the order slice.
//
// @lc code=start

import "sort"

func FindLUSlength522(strs []string) int {
	order := make([]int, len(strs))
	for i := range order {
		order[i] = i
	}
	sort.Slice(order, func(i, j int) bool {
		return len(strs[order[i]]) > len(strs[order[j]])
	})

	for _, index := range order {
		candidate := strs[index]
		uncommon := true
		for otherIndex, other := range strs {
			if otherIndex == index {
				continue
			}
			if len(other) >= len(candidate) && isSubsequence522(candidate, other) {
				uncommon = false
				break
			}
		}
		if uncommon {
			return len(candidate)
		}
	}
	return -1
}

func isSubsequence522(small string, large string) bool {
	pointer := 0
	for i := range large {
		if pointer < len(small) && small[pointer] == large[i] {
			pointer++
		}
	}
	return pointer == len(small)
}

// @lc code=end
